2996_reflection/reflection.md

```diff
@@ -2283,6 +2284,86 @@ Lastly, we clarify that during the definition of an _injected declaration_, the
 This machinery is "off in the weeds" of technicalities related to modules, lookup, etc., but we believe (hope?) that it provides a sound basis upon which to build generative reflection within the framework provided by core language wording: not only for P2996, but for future papers as well.
 
 
+### Restrictions on injected declarations
+
+The advancement of this proposal through WG21 has naturally led to increased scrutiny of the mechanisms here proposed. One such area is the possibility of leveraging injected declarations to observe failed template substitutions. Consider the following example:
+
+```cpp
+struct S;
+
+template <typename> struct TCls {
+  static consteval bool sfn()  // #1
+      requires ([] {
+        consteval {
+          define_aggregate(^^S, {});
+        }
+      }(), false) {
+    return false;  // never selected
+  }
+
+  static consteval bool sfn()  // #2
+      requires (true) {
+    return true;   // always selected
+  }
+};
+
+static_assert(TCls<void>::sfn());
+static_assert(is_complete_type(^^S));
+```
+
+The above example observes the effects of the failed substitution of `#1` by way of the completeness of `S`. Such tricks can be used to observe implementation details, like the order in which overloads are checked, that may be unportable (and which implementations might desire to change over time).
+
+Our proposed solution, specified in [expr.const]/23.2, is to make it ill-formed to produce an injected declaration _from_ a manifestly constant-evaluated expression within an instantiation to _outside of_ that instantiation, or visa versa. Because that expression in the example above (`define_aggregate(^^S, {})`) is within the instantiation of the requires clause of `TCls<void>::sfn`, and the target scope of the injected declaration is outside of that same instantiaton, the example becomes ill-formed (diagnostic required). Note that this does not prevent writing `consteval` function templates that wrap `define_aggregate`:
```

Comment:
I think better to explicitly say: "produce an injected declaration _from inside of_ a manifestly constant-evaluated expression within an instantiation to _outside of_ that instantiation". Specifically having both INSIDE and OUTSIDE italicized helps make that clear.

Reply:
Done.